packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/media/ImageStreamReader.java

```diff
@@ -0,0 +1,223 @@
+      if (dartImageFormat == ImageFormat.NV21) {
+        imageBuffer.put("planes", parsePlanesForNv21(image));
+      } else {
+        imageBuffer.put("planes", parsePlanesForYuvOrJpeg(image));
+      }
+
+      imageBuffer.put("width", image.getWidth());
+      imageBuffer.put("height", image.getHeight());
+      imageBuffer.put("format", dartImageFormat);
+      imageBuffer.put("lensAperture", captureProps.getLastLensAperture());
+      imageBuffer.put("sensorExposureTime", captureProps.getLastSensorExposureTime());
+      Integer sensorSensitivity = captureProps.getLastSensorSensitivity();
+      imageBuffer.put(
+          "sensorSensitivity", sensorSensitivity == null ? null : (double) sensorSensitivity);
+
+      final Handler handler = new Handler(Looper.getMainLooper());
+      handler.post(() -> imageStreamSink.success(imageBuffer));
+      image.close();
+
+    } catch (IllegalStateException e) {
+      // Handle "buffer is inaccessible" errors that can happen on some devices from ImageStreamReaderUtils.yuv420ThreePlanesToNV21()
+      final Handler handler = new Handler(Looper.getMainLooper());
+      handler.post(
+          () ->
+              imageStreamSink.error(
+                  "IllegalStateException",
+                  "Caught IllegalStateException: " + e.getMessage(),
+                  null));
+      image.close();
+    }
+  }
+
+  /**
+   * Given an input image, will return a list of maps suitable to send back to dart where
+   * each map describes the image plane.
+   *
+   * For Yuv / Jpeg, we do no further processing on the frame so we simply send it as-is.
+   *
+   * @param image - the image to process.
+   * @return parsed map describing the image planes to be sent to dart.
+   */
+  public List<Map<String, Object>> parsePlanesForYuvOrJpeg(@NonNull Image image) {
+    List<Map<String, Object>> planes = new ArrayList<>();
+
+    // For YUV420 and JPEG, just send the data as-is for each plane.
+    for (Image.Plane plane : image.getPlanes()) {
+      ByteBuffer buffer = plane.getBuffer();
+
+      byte[] bytes = new byte[buffer.remaining()];
+      buffer.get(bytes, 0, bytes.length);
+
+      Map<String, Object> planeBuffer = new HashMap<>();
+      planeBuffer.put("bytesPerRow", plane.getRowStride());
+      planeBuffer.put("bytesPerPixel", plane.getPixelStride());
+      planeBuffer.put("bytes", bytes);
+
+      planes.add(planeBuffer);
+    }
+    return planes;
+  }
+
+  /**
+   * Given an input image, will return a single-plane NV21 image. Assumes YUV420 as an input type.
+   *
+   * @param image - the image to process.
+   * @return parsed map describing the image planes to be sent to dart.
+   */
+  public List<Map<String, Object>> parsePlanesForNv21(@NonNull Image image) {
+    List<Map<String, Object>> planes = new ArrayList<>();
+
+    // We will convert the YUV data to NV21 which is a single-plane image
+    ByteBuffer bytes =
+        imageStreamReaderUtils.yuv420ThreePlanesToNV21(
```

Comment:
Why not catch the illegal state exception here? Just wondering if this may not be the only cause.

Reply:
I opted for doing it in the other way because I wanted to make sure we post an exception to dart and close the image after. If we catch it here then i'm not sure how we'd still return an error to dart because line 101 would presumedly return normally, which would then have the plugin try and post an image back to dart.